File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        int[] dp = new int[nums.length+1];
        Arrays.fill(dp,-1);

        return robbing(nums,0,dp);
    }
    public int robbing(int[] nums,int i,int[] dp){
        if(i>=nums.length){
            return 0;
        }
        if(dp[i]!=-1){
            return dp[i];
        }
        
        int inc=nums[i]+robbing(nums,i+2,dp);
        int ni=robbing(nums,i+1,dp);
     
        dp[i]=Math.max(ni,inc);
        return dp[i];
    }
}

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        vector<int> dp(nums.size()+1,-1); 
        return robbery(nums,0,dp);
    }
    int robbery(vector<int>& nums , int i,vector<int>& dp){

        if(i>=nums.size()){
            return 0;
        }
        if(dp[i]!=-1) return dp[i];
        int c1=nums[i]+robbery(nums,i+2,dp);
        int c2=robbery(nums,i+1,dp);
        return dp[i]=max(c1,c2);
    }
};